README.md
<p align="center">
  <img alt="Eat at EPFL" src="https://raw.githubusercontent.com/innovativeinnovation/eat-at-epfl/master/docs/readme/readme-logo.png">
</p>

<p align="center">
  What to eat at EPFL.
</p>

<p align="center">
  <a href="https://github.com/innovativeinnovation/eat-at-epfl/actions">
    <img alt="Build Status" src="https://github.com/innovativeinnovation/eat-at-epfl/workflows/Build/badge.svg?branch=master">
  </a>
  <a href="https://raw.githubusercontent.com/innovativeinnovation/eat-at-epfl/master/LICENSE">
    <img alt="Apache License 2.0" src="https://img.shields.io/badge/license-Apache%202.0-blue.svg">
  </a>
  <a href='https://www.npmjs.com/package/eat-at-epfl'>
    <img alt="NPM Version" src="https://img.shields.io/npm/v/eat-at-epfl.svg" />
  </a>
</p>

---

Install
-------

Install this globally and you'll have access to the `eat-at-epfl` command
anywhere on your system.

```bash
npm i eat-at-epfl -g
```

Usage
-----

```console
eat-at-epfl -h
Usage: eat-at-epfl [options]

Options:
  -a, --all               List all restaurants
  -d, --date              A date. Example: 18/04/2019            [string]
  -e, --evening           Evening menus
  -l, --language          Show menus in "en" or "fr"             [string]
  -p, --pick-random-menu  Pick a random menu
  -r, --restoId           Restaurant ID                          [number]
  -t, --tags              A comma separated list of menu types   [string]
  -h, --help              Show help                             [boolean]
  -v, --version           Show version number                   [boolean]

Examples:
  eat-at-epfl
  eat-at-epfl -e -l fr
  eat-at-epfl -t Pizza,Fish
  eat-at-epfl -d 04/05/2019
```

Screenshot
----------

![command line screenshot](https://raw.githubusercontent.com/innovativeinnovation/eat-at-epfl/master/docs/readme/screenshot.png)

See also
--------

* [epfl-menu-api](https://github.com/innovativeinnovation/epfl-menu-api)
* [epfl-menu](https://github.com/gcmalloc/epfl-menu)

Contributing
------------

Contributions are always welcome.

See [Contributing](CONTRIBUTING.md).

Developer
---------

* [William Belle](https://github.com/williambelle)

License
-------

Apache License 2.0

(c) William Belle, 2019-2021.

See the [LICENSE](LICENSE) file for more details.


CHANGELOG.md
Changelog
=========

### v3.0.0 / 2021-08-26

  - Drop support for Node < 12
  - Test against Node.js 16
  - Update dependencies

### v2.1.3 / 2021-04-09

  - Update and clean dependencies
  - Migrate from Travis CI to GitHub Actions
  - Remove badge dependencies

### v2.1.2 / 2020-11-17

  - Update dependencies
  - Test against Node.js 15
  - Migrate travis-ci.org to .com

### v2.1.1 / 2020-07-09

  - Update dependencies
  - Test against Node.js 14  

### v2.1.0 / 2020-02-08

  - Add option to pick a random menu
  - Update dependencies

### v2.0.0 / 2020-01-02

  - Drop support for Node < 10
  - Update dependencies
  - Allow list of tags in English

### v1.1.3 / 2019-12-14

  - Update dependencies
  - Test against Node.js 13

### v1.1.2 / 2019-10-06

  - Update dependencies

### v1.1.1 / 2019-07-20

  - Improve documentation
  - Update dependencies

### v1.1.0 / 2019-05-19

  - Add cli option `all`
  - Add tags translation
  - Display resto id

### v1.0.0 / 2019-05-16

  - Add cli option `date`
  - Add cli option `tags`
  - Improve tests
  - Update dependencies
  - Update screenshot

### v0.1.0 / 2019-05-07

  - Add restaurant address
  - Add cli option `language`
  - Add cli option `restoId`

### v0.0.1 / 2019-05-05

  - First version, released on an unsuspecting world.


test/test.js
/*
 * (c) William Belle, 2019-2021.
 * See the LICENSE file for more details.
 */

require('chai').should();
const version = require('../package').version;

describe('eat-at-epfl cli', function () {
  this.timeout(15000);
  let nextCliOption = ['-v'];
  let response;

  beforeEach((done) => {
    const execFile = require('child_process').execFile;
    execFile('./src/cli.js', nextCliOption, (error, stdout, stderr) => {
      if (error) {
        throw error;
      }
      response = stdout || stderr;
      done();
    });
  });

  it('should match version with option -v', () => {
    response.should.equal(version + '\n');
    nextCliOption = ['--date=30/02/2019'];
  });

  it('should match "Usage:" with option --date=30/02/2019', () => {
    response.should.match(/Usage:/);
    nextCliOption = ['--date=26/08/2021'];
  });

  it('should match "L\'Arcadie" with -d 26/08/2021', () => {
    response.should.match(/L'Arcadie/);
    nextCliOption = ['-e', '--date=26/08/2021'];
  });

  it('should match "Sushizen" with -d 26/08/2021 -e', () => {
    response.should.match(/Sushizen/);
    nextCliOption = ['-e', '--language=fr', '--date=25/08/2021'];
  });

  it('should match "teriyaki" with -d 25/08/2021 -e -l fr', () => {
    response.should.match(/teriyaki/);
    nextCliOption = ['--tags=Viande', '--date=26/08/2021'];
  });

  it('should match "Lamb chawarma" with -d 26/08/2021 -t Viande', () => {
    response.should.match(/Lamb chawarma/);
    nextCliOption = ['--restoId=14', '--date=25/08/2021'];
  });

  it('should match "Hong Thaï Rung" with -r 14 --date=25/08/2021', () => {
    response.should.match(/Hong Thaï Rung/);
    nextCliOption = ['--restoId=bo', '--date=18/04/2019'];
  });

  it('should match "Usage:" with option --date=18/04/2019 -r bo', () => {
    response.should.match(/Usage:/);
    nextCliOption = ['-a'];
  });

  it('should match "32. L\'Esplanade (CO160)" with option -a', () => {
    response.should.match(/32. L'Esplanade \(CO160\)/);
  });
});


src/cli.js
#!/usr/bin/env node

/*
 * (c) William Belle, 2019-2021.
 * See the LICENSE file for more details.
 */

const epflMenuApi = require('epfl-menu-api');
const moment = require('moment');
const chalk = require('chalk');
const yargs = require('yargs')

  // List resto
  .option('a', {
    alias: 'all',
    describe: 'List all restaurants'
  })

  // Date
  .option('d', {
    alias: 'date',
    describe: 'A date. Example: 18/04/2019',
    requiresArg: true,
    type: 'string'
  })

  // Evening
  .option('e', {
    alias: 'evening',
    describe: 'Evening menus'
  })

  // Language
  .option('l', {
    alias: 'language',
    describe: 'Show menus in "en" or "fr"',
    requiresArg: true,
    type: 'string'
  })

  .option('p', {
    alias: 'pick-random-menu',
    describe: 'Pick a random menu'
  })

  // Resto ID
  .option('r', {
    alias: 'restoId',
    describe: 'Restaurant ID',
    requiresArg: true,
    type: 'number'
  })

  // Tags
  .option('t', {
    alias: 'tags',
    describe: 'A comma separated list of menu types',
    requiresArg: true,
    type: 'string'
  })

  // Version
  .alias('v', 'version')

  // Help
  .help('h')
  .alias('h', 'help')
  .usage('Usage: $0 [options]')
  .example('$0')
  .example('$0 -e -l fr')
  .example('$0 -t Pizza,Fish')
  .example('$0 -d 04/05/2019');

const argv = yargs.argv;
const opts = {};

if (argv.a) {
  epflMenuApi.findResto().then(listRestos => {
    listRestos.sort(sortRestoByName);
    putListResto(listRestos);
  });
} else {
  opts.language = 'en';
  if (argv.e) {
    opts.partOfDay = 'soir';
  }
  if (argv.l && argv.l === 'fr') {
    opts.language = 'fr';
  }
  if (isNaN(argv.r) && argv.r !== undefined) {
    yargs.showHelp();
    process.exit(0);
  } else {
    if (argv.r !== undefined) {
      opts.restoId = argv.r;
    }
  }
  if (argv.t) {
    opts.tags = argv.t;
  }
  if (argv.d) {
    if (moment(argv.d, 'DD/MM/YYYY', true).isValid()) {
      opts.date = argv.d;
    } else {
      yargs.showHelp();
      process.exit(0);
    }
  }

  const jsonRestos = epflMenuApi.findResto();
  const jsonMenus = epflMenuApi.findMenu(opts);

  Promise.all([jsonRestos, jsonMenus]).then(values => {
    let menuList = values[1];
    if (argv.p && menuList.length > 0) {
      menuList = getRamdomMenu(values[1]);
    }
    const listRestoWithPlan = buildListRestoWithPlan(values[0]);
    const listRestoWithListMenu = buildListRestoWithListMenu(menuList);
    put(listRestoWithPlan, listRestoWithListMenu);
  }).catch(function (err) {
    console.log(err);
  });
}

const getRamdomMenu = (list) => {
  const menu = list[Math.floor(Math.random() * list.length)];
  return [menu];
};

const putListResto = (listResto) => {
  for (let i = 0; i < listResto.length; i++) {
    console.log(chalk.blue(buildRestoLine(
      listResto[i].restoName,
      listResto[i].plan,
      listResto[i].restoID
    )));
  }
};

const sortRestoByName = (a, b) => {
  const nameA = a.restoName.toUpperCase();
  const nameB = b.restoName.toUpperCase();
  if (nameA < nameB) {
    return -1;
  }
  if (nameA > nameB) {
    return 1;
  }
  return 0;
};

const buildListRestoWithPlan = (jsonRestos) => {
  const listRestoWithPlan = [];
  for (let i = 0; i < jsonRestos.length; i++) {
    const restoName = jsonRestos[i].restoName;
    listRestoWithPlan[restoName] = {
      location: jsonRestos[i].plan,
      id: jsonRestos[i].restoID
    };
  }
  return listRestoWithPlan;
};

const buildListRestoWithListMenu = (jsonMenus) => {
  const listRestoWithListMenu = [];
  for (let i = 0; i < jsonMenus.length; i++) {
    const restoName = jsonMenus[i].restoName;
    if (Array.isArray(listRestoWithListMenu[restoName])) {
      listRestoWithListMenu[restoName].push(jsonMenus[i]);
    } else {
      listRestoWithListMenu[restoName] = [];
      listRestoWithListMenu[restoName].push(jsonMenus[i]);
    }
  }
  return listRestoWithListMenu;
};

const buildRestoLine = (name, location, id) => {
  let line = id + '. ' + name;
  if (location !== '') {
    line += ' (' + location + ')';
  }
  return line;
};

const BuildMenuLines = (name, pos, listMenu) => {
  let line = '';
  const menuPart = ['platPrincipal', 'accompFeculents', 'accompLegumes'];
  for (let k = 0; k < menuPart.length; k++) {
    if (listMenu[name][pos][menuPart[k]] !== '') {
      if (menuPart[k] === 'platPrincipal') {
        line += '    🍽  ';
      } else {
        line += '       ';
      }
      line += listMenu[name][pos][menuPart[k]] + '\n';
    }
  }
  return line;
};

const BuildTagsLine = (name, pos, listMenu) => {
  let line = '';
  if (listMenu[name][pos].menuTags !== '') {
    let tags = listMenu[name][pos].menuTags;
    if (argv.l !== 'fr') {
      tags = epflMenuApi.translateTags(tags);
    }
    line += '      [Tags: ' + tags + ']\n';
  }
  return line;
};

const put = (listResto, listMenu) => {
  for (const key in listMenu) {
    console.log(chalk.blue(buildRestoLine(
      key, listResto[key].location, listResto[key].id
    )));
    for (let j = 0; j < listMenu[key].length; j++) {
      console.log(
        chalk.green(BuildMenuLines(key, j, listMenu)),
        chalk.yellow(BuildTagsLine(key, j, listMenu))
      );
    }
  }
};
